File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/CompleteCheckpointLogRecord.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System.Diagnostics.CodeAnalysis;
    using System.Fabric;
    using System.IO;

    internal sealed class CompleteCheckpointLogRecord : LogHeadRecord
    {
        private static readonly uint ApproximateDiskSpaceUsed = sizeof(int);

        [SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1126:PrefixCallsCorrectly", Justification = "Reviewed.")]
        internal CompleteCheckpointLogRecord(LogRecordType recordType, ulong recordPosition, long lsn)
            : base(recordType, recordPosition, lsn)
        {
            Utility.Assert(
                recordType == LogRecordType.CompleteCheckpoint,
                "Record type is expected to be end rename but the record type is : {0}",
                recordType);
        }

        internal CompleteCheckpointLogRecord(
            LogicalSequenceNumber lsn,
            IndexingLogRecord logHeadRecord,
            PhysicalLogRecord lastLinkedPhysicalRecord)
            : base(LogRecordType.CompleteCheckpoint, logHeadRecord, lsn, lastLinkedPhysicalRecord)
        {
        }

        protected override void Read(BinaryReader br, bool isPhysicalRead)
        {
            base.Read(br, isPhysicalRead);
            this.ReadPrivate(br, isPhysicalRead);

            return;
        }

        protected override void ReadLogical(OperationData operationData, ref int index)
        {
            base.ReadLogical(operationData, ref index);

            using (var br = new BinaryReader(IncrementIndexAndGetMemoryStreamAt(operationData, ref index)))
            {
                this.ReadPrivate(br, false);
            }

            return;
        }

        protected override void Write(BinaryWriter bw, OperationData operationData, bool isPhysicalWrite, bool forceRecomputeOffsets)
        {
            base.Write(bw, operationData, isPhysicalWrite, forceRecomputeOffsets);
            var startingPosition = bw.BaseStream.Position;

            // Metadata Size
            bw.BaseStream.Position += sizeof(int);

            // Future fields

            // End of metadata.
            var endPosition = bw.BaseStream.Position;
            var sizeOfSection = checked((int) (endPosition - startingPosition));
            bw.BaseStream.Position = startingPosition;
            bw.Write(sizeOfSection);
            bw.BaseStream.Position = endPosition;

            operationData.Add(CreateArraySegment(startingPosition, bw));
        }

        private void ReadPrivate(BinaryReader br, bool isPhysicalRead)
        {
            // Read Metadata size.
            var logicalStartingPosition = br.BaseStream.Position;
            var logicalSizeOfSection = br.ReadInt32();
            var logicalEndPosition = logicalStartingPosition + logicalSizeOfSection;

            // Read Metadata fields (future)

            // Jump to the end of the section ignoring fields that are not understood.
            Utility.Assert(logicalEndPosition >= br.BaseStream.Position, "Could not have read more than section size.");
            br.BaseStream.Position = logicalEndPosition;

            this.UpdateApproximateDiskSize();
        }

        private void UpdateApproximateDiskSize()
        {
            this.ApproximateSizeOnDisk += ApproximateDiskSpaceUsed;
        }
    }
}

File: src/prod/src/data/txnreplicator/logrecordlib/CompleteCheckpointLogRecord.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

const ULONG CompleteCheckPointLogRecord::DiskSpaceUsed = sizeof(LONG32);

CompleteCheckPointLogRecord::CompleteCheckPointLogRecord(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in IndexingLogRecord & invalidIndexingLogRecord)
    : LogHeadRecord(recordType, recordPosition, lsn, invalidPhysicalLogRecord, invalidIndexingLogRecord)
{
}

CompleteCheckPointLogRecord::CompleteCheckPointLogRecord(
    __in LONG64 lsn,
    __in IndexingLogRecord & logHeadRecord,
    __in_opt PhysicalLogRecord * const lastLinkedPhysicalRecord,
    __in PhysicalLogRecord & invalidPhysicalLogRecord)
    : LogHeadRecord(LogRecordType::Enum::CompleteCheckpoint, logHeadRecord, lsn, lastLinkedPhysicalRecord, invalidPhysicalLogRecord)
{
    UpdateApproximateDiskSize();
}

CompleteCheckPointLogRecord::~CompleteCheckPointLogRecord()
{
}

CompleteCheckPointLogRecord::SPtr CompleteCheckPointLogRecord::Create(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in IndexingLogRecord & invalidIndexingLogRecord,
    __in KAllocator & allocator)
{
    CompleteCheckPointLogRecord* pointer = _new(COMPLETECHECKPOINTLOGRECORD_TAG, allocator) CompleteCheckPointLogRecord(
        recordType, 
        recordPosition, 
        lsn, 
        invalidPhysicalLogRecord, 
        invalidIndexingLogRecord);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return CompleteCheckPointLogRecord::SPtr(pointer);
}

CompleteCheckPointLogRecord::SPtr CompleteCheckPointLogRecord::Create(
    __in LONG64 lsn,
    __in IndexingLogRecord & logHeadRecord,
    __in_opt PhysicalLogRecord * const lastLinkedPhysicalRecord,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in KAllocator & allocator)
{
    CompleteCheckPointLogRecord* pointer = _new(COMPLETECHECKPOINTLOGRECORD_TAG, allocator) CompleteCheckPointLogRecord(
        lsn, 
        logHeadRecord, 
        lastLinkedPhysicalRecord, 
        invalidPhysicalLogRecord);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return CompleteCheckPointLogRecord::SPtr(pointer);
}

std::wstring CompleteCheckPointLogRecord::ToString() const
{
    std::wstring logRecordString = Constants::CloseJSON;

    return __super::ToString() + logRecordString;
}

void CompleteCheckPointLogRecord::UpdateApproximateDiskSize()
{
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + DiskSpaceUsed;
}

void CompleteCheckPointLogRecord::Read(
    __in BinaryReader & binaryReader,
    __in bool isPhysicalRead)
{
    __super::Read(binaryReader, isPhysicalRead);

    // Read Metadata size
    ULONG32 logicalStartingPosition = binaryReader.Position;
    ULONG32 logicalSizeOfSection = 0;
    binaryReader.Read(logicalSizeOfSection);
    ULONG32 logicalEndPosition = logicalStartingPosition + logicalSizeOfSection;

    ASSERT_IFNOT(
        logicalEndPosition >= binaryReader.Position,
        "Invalid complete checkpoint log record, logical end position: {0} reader position:{1}", 
        logicalEndPosition, binaryReader.Position);

    binaryReader.Position = logicalEndPosition;

    UpdateApproximateDiskSize();
}

void CompleteCheckPointLogRecord::Write(
    __in BinaryWriter & binaryWriter,
    __inout OperationData & operationData,
    __in bool isPhysicalWrite,
    __in bool forceRecomputeOffsets)
{
    __super::Write(binaryWriter, operationData, isPhysicalWrite, forceRecomputeOffsets);
    ULONG32 startingPosition = binaryWriter.Position;

    // Metadata size
    binaryWriter.Position += sizeof(ULONG32);

    // Future fields

    // End of Metadata
    ULONG32 endPosition = binaryWriter.Position;
    ULONG32 sizeOfSection = endPosition - startingPosition;

    binaryWriter.Position = startingPosition;
    binaryWriter.Write(sizeOfSection);
    binaryWriter.Position = endPosition;

    KBuffer::SPtr kBuffer = binaryWriter.GetBuffer(startingPosition);
    operationData.Append(*kBuffer);
}
